File: 1851-maximum-number-of-events-that-can-be-attended-ii/1851-maximum-number-of-events-that-can-be-attended-ii.cpp
class Solution {
public:
    int maxValue(vector<vector<int>>& events, int k) {
        int n = events.size();
        sort(events.begin(), events.end());
        auto bs = [&](int target) -> int {
            int l = 0, r = n;
            while(l < r){
                int mid = l + (r - l) / 2;
                if(events[mid][0] <= target) l = mid + 1;
                else r = mid;
            }
            return l;
        };
        vector<vector<int>> dp(n + 1, vector<int> (k + 1, 0));

        // got TLE on recursive approach with memo, gotta tabulate now

        // function<int(int,int)> helper = [&](int i, int j) -> int {
        //     if(j == k || i == n) return 0;
        //     if(dp[i][j] != -1) return dp[i][j];
        //     else return dp[i][j] = max(events[i][2] + helper(bs(events[i][1]), j + 1), helper(i + 1, j));
        // };

        // Jst realized i can store all next as precomp instead of bs everytime

        vector<int> next(n, 0);
        for(int i = 0; i<n; i++) next[i] = bs(events[i][1]);

        for(int i = n - 1; i>=0; i--){
            for(int j = k - 1; j>=0; j--){
                dp[i][j] = dp[i + 1][j];
                dp[i][j] = max(dp[i][j], events[i][2] + dp[next[i]][j + 1]);
            }
        }
        return dp[0][0];
    }
};

File: 1851-maximum-number-of-events-that-can-be-attended-ii/1851-maximum-number-of-events-that-can-be-attended-ii.go
func maxValue(events [][]int, k int) int {
    slices.SortFunc(events, func(a, b []int) int {
        return a[0] - b[0]
    })
    n := len(events)
    bs := func (target int) int {
        l, r := 0, n
        for l < r {
            mid := l + (r - l) / 2
            if events[mid][0] <= target {
                l = mid + 1
            } else {
                r = mid
            }
        }
        return l
    }
    next := make([]int, n)
    for i := 0; i<n; i++ {
        next[i] = bs(events[i][1])
    }
    dp := make([][]int, n + 1)
    for i := 0; i<=n; i++ {
        dp[i] = make([]int, k + 1)
    }
    for i := n - 1; i>=0; i-- {
        for j := k - 1; j>=0; j-- {
            dp[i][j] = max(dp[i + 1][j], events[i][2] + dp[next[i]][j+1]);
        }
    }
    return dp[0][0]
}

func max(a, b int) int {
    if a > b {
        return a
    } else {
        return b
    }
}